src/Lib.P2P/SecureCommunication/Secio1.cs

```diff
@@ -0,0 +1,246 @@
+                ms.Position = 0;
+                oh1 = hasher.ComputeHash(ms);
+            }
+
+            using (var hasher = MultiHash.GetHashAlgorithm())
+            await using (var ms = new MemoryStream())
+            {
+                ms.Write(localProposal.PublicKey, 0, localProposal.PublicKey.Length);
+                ms.Write(remoteProposal.Nonce, 0, remoteProposal.Nonce.Length);
+                ms.Position = 0;
+                oh2 = hasher.ComputeHash(ms);
+            }
+
+            var order = 0;
+            for (var i = 0; order == 0 && i < oh1.Length; ++i) order = oh1[i].CompareTo(oh2[i]);
+            if (order == 0)
+            {
+                throw new Exception("Same keys and nonces; talking to self");
+            }
+            
+            var curveName = SelectBest(order, localProposal.Exchanges, remoteProposal.Exchanges);
+            if (curveName == null)
+            {
+                throw new Exception("Cannot agree on a key exchange.");
+            }
+
+            var cipherName = SelectBest(order, localProposal.Ciphers, remoteProposal.Ciphers);
+            if (cipherName == null)
+            {
+                throw new Exception("Cannot agree on a chipher.");
+            }
+
+            var hashName = SelectBest(order, localProposal.Hashes, remoteProposal.Hashes);
+            if (hashName == null)
+            {
+                throw new Exception("Cannot agree on a hash.");
+            }
+
+            // =============================================================================
+            // step 2. Exchange -- exchange (signed) ephemeral keys. verify signatures.
+
+            // Generate EphemeralPubKey
+            var localEphemeralKey = EphermalKey.Generate(curveName);
+            var localEphemeralPublicKey = localEphemeralKey.PublicKeyBytes();
+
+            // Send Exchange packet
+            var localExchange = new Secio1Exchange();
+            await using (var ms = new MemoryStream())
+            {
+                Serializer.Serialize(ms, localProposal);
+                Serializer.Serialize(ms, remoteProposal);
+                ms.Write(localEphemeralPublicKey, 0, localEphemeralPublicKey.Length);
+                localExchange.Signature = connection.LocalPeerKey.Sign(ms.ToArray());
+            }
+
+            localExchange.EPublicKey = localEphemeralPublicKey;
+            Serializer.SerializeWithLengthPrefix(stream, localExchange, PrefixStyle.Fixed32BigEndian);
+            await stream.FlushAsync(cancel).ConfigureAwait(false);
+
+            // Receive their Exchange packet.  If nothing, then most likely the
+            // remote has closed the connection because it does not like us.
+            var remoteExchange =
+                Serializer.DeserializeWithLengthPrefix<Secio1Exchange>(stream, PrefixStyle.Fixed32BigEndian);
+            if (remoteExchange == null)
+            {
+                throw new Exception("Remote refuses the SECIO exchange.");
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['System.Exception' should not be thrown by user code.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)